all-the-icons-dired.el

```diff
@@ -57,92 +56,49 @@
-              (all-the-icons-dired--add-overlay (dired-move-to-filename) (concat icon "\t"))))))
-      (forward-line 1))))
-
-(defun all-the-icons-dired--refresh-advice (fn &rest args)
-  "Advice function for FN with ARGS."
-  (prog1 (apply fn args)
-    (when all-the-icons-dired-mode
-      (all-the-icons-dired--refresh))))
-
-(defvar all-the-icons-dired-advice-alist
-  '((dired-aux     dired-do-redisplay           all-the-icons-dired--refresh-advice)
-    (dired-aux     dired-create-directory       all-the-icons-dired--refresh-advice)
-    (dired-aux     dired-do-create-files        all-the-icons-dired--refresh-advice)
-    (dired-aux     dired-do-kill-lines          all-the-icons-dired--refresh-advice)
-    (dired-aux     dired-do-rename              all-the-icons-dired--refresh-advice)
-    (dired-aux     dired-insert-subdir          all-the-icons-dired--refresh-advice)
-    (dired-aux     dired-kill-subdir            all-the-icons-dired--refresh-advice)
-    (dired         wdired-abort-changes         all-the-icons-dired--refresh-advice)
-    (dired         dired-internal-do-deletions  all-the-icons-dired--refresh-advice)
-    (dired-narrow  dired-narrow--internal       all-the-icons-dired--refresh-advice)
-    (dired-subtree dired-subtree-insert         all-the-icons-dired--refresh-advice)
-    (dired-subtree dired-subtree-remove         all-the-icons-dired--refresh-advice)
-    (dired         dired-readin                 all-the-icons-dired--refresh-advice)
-    (dired         dired-revert                 all-the-icons-dired--refresh-advice)
-    (find-dired    find-dired-sentinel          all-the-icons-dired--refresh-advice))
-  "A list of file, adviced function, and advice function.")
+(defun all-the-icons-dired--icon (file)
+  "Return the icon for FILE."
+  (if (file-directory-p file)
+      (all-the-icons-icon-for-dir file
+                                  :face 'all-the-icons-dired-dir-face
+                                  :v-adjust all-the-icons-dired-v-adjust)
+    (apply 'all-the-icons-icon-for-file file
+           (append
+            `(:v-adjust ,all-the-icons-dired-v-adjust)
+            (when all-the-icons-dired-monochrome
+              `(:face ,(face-at-point)))))))
+
+(defun all-the-icons-dired--put-icon (pos) "Propertize POS with icon."
+       (let* ((file (dired-get-filename 'relative 'noerror))
+              (icon (all-the-icons-dired--icon file)))
+         (put-text-property (1- pos) pos 'display
+                            (if (member file '("." ".."))
+                                "    "
+                              (concat " " icon " ")))))
+
+(defun all-the-icons-dired--propertize (&optional beg end &rest _)
+  "Add icons using text properties from BEG to END.
+They defualt to `(point-min)' and `(point-max)'."
+  (let ((beg (or beg (point-min)))
+        (end (or end (point-max))))
+    (with-silent-modifications
+      (ignore-errors
```

Comment:
I had copied this from the `dired-insert-set-properties` which has the same `ignore-errors` around the call to `dired-move-to-filename` when I encountered an error and it resolved the error and I didn't look further. Poking around the stack trace now, it seems this is due to `dired-subdir-alist` not being set when `all-the-icons--propertize` is called. The error is actually due to the `(dired-get-filename 'relative 'noerror)` inside  `all-the-icons-dired--put-icon`. `dired-get-filename` calls `dired-current-directory` which is what raises the error.

I think this can be fixed by changing the `(dired-get-filename 'relative 'noerror)` to `(dired-get-filename 'verbatim 'noerror)` since the filename as it appears in the buffer should be enough to decide which icon to put. Should I do that or do you see a reason to keep the first argument as `relative`? I have tested the `verbatim` options and it seems to work fine.

Reply:
The old code guards this when-let block with 

```elisp
(when (dired-move-to-filename nil)
        (let ((case-fold-search t))
```

What's the reason for removing them? Could this be the cause of your errors?

Reply:
> The old code guards this when-let block with
> 
> ```emacs-lisp
> (when (dired-move-to-filename nil)
>         (let ((case-fold-search t))
> ```
> 
> What's the reason for removing them? Could this be the cause of your errors?

No, the cause was the advise to `dired-insert-set-properties` which the old code didn't have. The error arise because in the code for `(dired-mode)`, `dired-subdir-alist` is set after `dired-readin` and `dired-subdir-alist` is needed to compute file names relative to current directory. 

But I got rid of the errors, see the EDIT to my earlier comment  https://github.com/wyuenho/all-the-icons-dired/pull/17#discussion_r901851630

Reply:
Another way of dealing with the same issue would be to use `dired-re-dir` to match the directory names in `all-the-icons-dired--put-icons` and treat the symbolic icons specially and call `file-directory-p` on their destination.

This is a little messy but has the advantage that we can remove the hook in `dired-after-readin-hook` and do everything with just the advice to `dired-insert-set-properties`.

Reply:
I don't understand under what circumstances `dired-subdir-alist` will be set to `nil` in the PR. It's only set to nil temporarily *after* `dired-readin-insert` is called, which will call `dired-insert-set-properties`.

Reply:
Oh I see, when the dired buffer is first read in, dired-subdir-alist will be nil. Makes sense.